### PR TITLE
Refactor Graql query language library - part 15

### PR DESCRIPTION
**Pull request by haikalpribadi**
# Why is this PR needed?

This PR is a continuation of PR #4646, and incremental work for issue #4456 where you can find the full description of the goal. In summary, we are refactoring Graql to extract the Graql language library from Grakn's implementation of Graql, and we start with simplifying the Graql codebase first.

# What does the PR do?

Graql Predicates (based on how we implemented it) are operations we do on an attribute. For example:
- `$x 18 isa age`: which includes an assignment predicate of `val 18` to the variable `$x`
- `$x >= 20`: which includes GTE (greater-than-or-equals) comparison `>= 18` on the  variable `$x`
- `$x has name contains "Bob"`: which includes a String check `contains "Bob"` on the invisible variable which will become the `name` of `$x`

This PR focuses on rewriting Graql Predicates, such that:
1. We rename `Predicates` to something else that does not clash with reasoner's terminology. We've picked `Operation`, which is rather generic but satisfactory in the current case given that the class is nested scoped as a nested class: `ValueProperty.Operation`
2. We differentiate two different usages of predicates: `Operation.Assignment` and `Operation.Comparison`
3. The language and execution data structure are separated:
    - `ValueProperty.Operation` represents the language data structure
    - `ValueExecutor.Operation` represents the execution data structure
4. The types of literal values that can be provided into a predicate/operation are strictly checked at compile time, by enumerating all possible attribute operations in the `AttributeProperties` interface in which `Statements` will implement. This seems rather verbose, but highly valuable as it makes the language more reliable and intuitive to use.
5. The types of literal values that can be stored in the data structure are also strictly guaranteed by modelling it into the `Generic` types of the data structure, as seen in `ValueProperty.Operation` and `ValueExecutor`

### CODE REVIEW

File: server/src/graql/query/Query.java [hunk 1]
```diff
@@ -95,7 +95,30 @@ public String toString() {
...
+            for (Operator c : values()) {
+                if (c.operator.equals(value)) {
+                    return c;
+                }
+            }
+            return null;
+        }
+    }
+
+    enum Comparator {
```
Comment (kasper-piskorski) [inline server/src/graql/query/Query.java]: why don't we make it a class? we would avoid all the procedural code and multiple ifs in ValueExecutor

File: server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java [hunk 2]
```diff
@@ -18,65 +18,104 @@
...
- *
- * <p>
  * Predicate implementation specialising it to be an id predicate. Corresponds to {@link IdProperty}.
- * </p>
- *
- *
  */
-@AutoValue
-public abstract class IdPredicate extends Predicate<ConceptId>{
+public class IdPredicate extends Predicate<ConceptId> {
```
Comment (kasper-piskorski) [inline server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java]: Still think it's a perfect autovalue use case here. Instead we are removing one line and adding forty for no gain.

File: server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java [hunk 3]
```diff
@@ -18,65 +18,104 @@
...
+
+    private final Variable varName;
```
Comment (kasper-piskorski) [inline server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java]: Without autovalue, these should go into the parent Predicate class.

File: server/src/graql/internal/reasoner/atom/predicate/NeqPredicate.java [hunk 4]
```diff
@@ -18,39 +18,73 @@
...
  * Predicate implementation specialising it to be an inequality predicate. Corresponds to graql {@link NeqProperty}.
- * </p>
- *
- *
  */
-@AutoValue
-public abstract class NeqPredicate extends Predicate<Variable> {
+public class NeqPredicate extends Predicate<Variable> {
+
+    private final Variable varName;
```
Comment (kasper-piskorski) [inline server/src/graql/internal/reasoner/atom/predicate/NeqPredicate.java]: ditto

File: server/src/graql/internal/reasoner/atom/predicate/ValuePredicate.java [hunk 5]
```diff
@@ -18,60 +18,94 @@
...
- *
- *
  */
-@AutoValue
-public abstract class ValuePredicate extends Predicate<grakn.core.graql.query.predicate.ValuePredicate> {
+public class ValuePredicate extends Predicate<ValueProperty.Operation<?>> {
 
-    @Override public abstract Statement getPattern();
-    @Override public abstract ReasonerQuery getParentQuery();
+    private final Variable varName;
```
Comment (kasper-piskorski) [inline server/src/graql/internal/reasoner/atom/predicate/ValuePredicate.java]: ditto

File: server/src/graql/query/Query.java (hunk 1 above)
Comment (haikalpribadi) [reply server/src/graql/query/Query.java]: The actual comparison class is in `(ValueProperty|ValueExecutor).Operation.Comparison`. These comparators you're referring to is just the syntax token.

File: server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java (hunk 3 above)
Comment (haikalpribadi) [reply server/src/graql/internal/reasoner/atom/predicate/IdPredicate.java]: Yes, that makes sense!

**Pull request merged**